### PR TITLE
V5 symmetric encryption

### PR DESCRIPTION
**Pull request by Mandragorian**
To enable stored messages to be protected using symmetric key encryption and validated using message authentication codes, this set of changes adds support for storing symmetric key material as Secret Key Packets, symmetric key encrypted session keys as Public Key Encrypted Session Key Packets, and MAC tags as Signature Packets.

(original PR at https://github.com/openpgpjs/openpgpjs/pull/1279 )

### CODE REVIEW

File: src/packet/public_key.js [hunk 1]
```diff
@@ -268,8 +268,10 @@ class PublicKeyPacket {
     const modulo = this.publicParams.n || this.publicParams.p;
     if (modulo) {
       result.bits = util.uint8ArrayBitLength(modulo);
-    } else {
+    } else if (this.publicParams.oid) {
       result.curve = this.publicParams.oid.getName();
+    } else {
+      result.symmetric = 'aes256';//this.publicParams.cipher.getName();
```
Comment (twiss) [inline src/packet/public_key.js]: This should probably the thing in the comment?

File: src/type/enum.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+ * Wrapper for enums
+ *
+ * @requires enums
+ * @module type/enum.js
+ */
+
+
+import enums from "../enums";
+
+const type_enum = e => class EnumType {
```
Comment (twiss) [inline src/type/enum.js]: This is a bit nitpicky but `e` is a variable name is usually short for `error` - even though I see that apparently enums.js uses `e` as an abbreviation too, but there it refers to the enum _value_, not the enum type. To avoid all confusion I would name it something more explicit here, e.g. `type`?

File: src/openpgp.js [hunk 3]
```diff
@@ -45,21 +45,22 @@ import util from './util';
...
+export async function generateKey({ userIDs = [], passphrase = "", type = "ecc", rsaBits = 4096, curve = "curve25519", keyExpirationTime = 0, date = new Date(), subkeys = [{}], config, symmetricHash = "sha256", symmetricCipher = "aes256" }) {
   config = { ...defaultConfig, ...config };
   userIDs = toArray(userIDs);
   if (userIDs.length === 0) {
     throw new Error('UserIDs are required for key generation');
   }
-  if (type === "rsa" && rsaBits < config.minRSABits) {
-    throw new Error(`rsaBits should be at least ${config.minRSABits}, got: ${rsaBits}`);
+  if (type === "rsa" && rsaBits < config.minRsaBits) {
+    throw new Error(`rsaBits should be at least ${config.minRsaBits}, got: ${rsaBits}`);
```
Comment (twiss) [inline src/openpgp.js]: This is a rebasing error, I think

File: src/packet/public_key_encrypted_session_key.js [hunk 4]
```diff
@@ -105,8 +105,9 @@ class PublicKeyEncryptedSessionKeyPacket {
       util.writeChecksum(this.sessionKey)
     ]);
     const algo = enums.write(enums.publicKey, this.publicKeyAlgorithm);
+    const privateParams = algo === enums.publicKey.aead ? key.privateParams : {};
```
Comment (twiss) [inline src/packet/public_key_encrypted_session_key.js]: Rather than passing an empty object, I think it might be good to pass `null` if possible, so that any attempt to access a private parameter in one of the other algorithms immediately fails. In addition, it should be slightly more efficient, since this object then doesn't have to be constructed. (Similarly everywhere else.)

File: src/type/short_byte_string.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    }
+    if (!util.isUint8Array(data)) {
+      throw new Error("data must be in the form of a Uint8Array");
+    }
+    this.data = data;
+    this.length = this.data.byteLength;
+  }
+
+  write() {
+    return util.concatUint8Array([util.writeNumber(this.length, 2), this.data]);
```
Comment (twiss) [inline src/type/short_byte_string.js]: I think we had decided in the end to do a single byte of length, right?

File: src/openpgp.js (hunk 3 above)
Comment (Mandragorian) [reply src/openpgp.js]: What is the error exactly? Should these lines be removed altogether?

Comment (twiss) [reply src/openpgp.js]: The patch is changing `minRSABits` to `minRsaBits`, undoing a change that was made somewhere along the lines in v5; these lines should not be changed here.

File: src/type/short_byte_string.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    }
+    if (!util.isUint8Array(data)) {
+      throw new Error("data must be in the form of a Uint8Array");
+    }
+    this.data = data;
+    this.length = this.data.byteLength;
+  }
+
+  write() {
+    return util.concatUint8Array([util.writeNumber(this.length, 1), this.data]);
```
Comment (twiss) [inline src/type/short_byte_string.js]:
```suggestion
    return util.concatUint8Array([new Uint8Array([this.length]), this.data]);
```

File: src/type/short_byte_string.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  }
+
+  read(input) {
+    if (input.length >= 1) {
+      const length = util.readNumber(input.subarray(0,1));
```
Comment (twiss) [inline src/type/short_byte_string.js]:
```suggestion
      const length = input[0];
```

File: src/type/short_byte_string.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+      if (input.length >= length + 1) {
+        this.data = input.subarray(1, 1 + length);
+        this.length = this.data.length;
```
Comment (twiss) [inline src/type/short_byte_string.js]:
```suggestion
        this.length = length;
```

File: test/crypto/crypto.js [hunk 9]
```diff
@@ -215,7 +215,7 @@ module.exports = () => describe('API functional testing', function() {
         openpgp.enums.publicKey.rsaEncryptSign, openpgp.enums.hash.sha1, RSAPublicParams, RSAPrivateParams, data, await crypto.hash.digest(2, data)
       );
       const success = await crypto.signature.verify(
-        openpgp.enums.publicKey.rsaEncryptSign, openpgp.enums.hash.sha1, RSAsignedData, RSAPublicParams, data, await crypto.hash.digest(2, data)
+        openpgp.enums.publicKey.rsaEncryptSign, openpgp.enums.hash.sha1, RSAsignedData, RSAPublicParams, {}, data, await crypto.hash.digest(2, data)
```
Comment (twiss) [inline test/crypto/crypto.js]: Also here in the tests, I think it would make sense to pass `null`

File: test/general/packet.js [hunk 10]
```diff
@@ -972,4 +972,102 @@ V+HOQJQxXJkVRYa3QrFUehiMzTeqqMdgC6ZqJy7+
...
     });
   });
+
+  it('Public key encrypted symmetric key packet using symmetric keys', () => {
+    const serializedPacket = new Uint8Array([
+      0x03, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+      0x00, 0x64, 0x01, 0x4c, 0x80, 0x3d, 0x9b, 0xb6,
+      0x44, 0x13, 0x25, 0x90, 0x24, 0x6e, 0x59, 0x0d,
+      0x68, 0xee, 0x1e, 0x23, 0x75, 0x8a, 0x90,
+      0x85, 0x78, 0x1b, 0xf7, 0xc4, 0x4d, 0x58, 0xc7,
```
Comment (twiss) [inline test/general/packet.js]: This is very nitpicky but maybe reformat this packet? It looks a bit strange now :sweat_smile:

Comment (Mandragorian) [reply test/general/packet.js]: Yeah I thought about this. I ended up leaving it as is since it keeps the data aligned the same way as in `expectedCipherText`. But yeah I guess it looks too weird.

Comment (twiss) [reply test/general/packet.js]: Oh, yeah, I didn't notice that. I guess that one was copied from this one, so if you want, you could reformat that one too :shrug:

File: src/crypto/crypto.js [hunk 11]
```diff
@@ -28,23 +28,29 @@ import publicKey from './public_key';
...
+import ShortByteString from '../type/short_byte_string';
+import hash from './hash';
+import config from '../config';
 import KDFParams from '../type/kdf_params';
+import { SymAlgoEnum, AEADEnum, HashEnum } from '../type/enum';
 import enums from '../enums';
 import util from '../util';
 import OID from '../type/oid';
 import { Curve } from './public_key/elliptic/curves';
+import crypto from './';
```
Comment (twiss) [inline src/crypto/crypto.js]: Rather than importing the entire crypto module, please import the specific submodule you need (e.g. `./mode`)

File: src/crypto/hmac.js [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+function HMAC(hmac) {
+  this.hmac = hmac;
+  this.update = function(data) {
+    this.hmac.process(data);
+  };
+  this.finalize = function() {
+    return this.hmac.finish().result;
+  };
+}
```
Comment (twiss) [inline src/crypto/hmac.js]: It seems like in all cases when we compute an hmac, we compute it in one go (not streaming), right? So I think it would be sufficient to expose a single function that computes an hmac in one go, rather than replicating an API similar to asmcrypto's.

File: src/crypto/hmac.js [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+ * @fileoverview This module implements an abstracted interface over
+ * HMAC implementation from asmcrypto.js
+ * @module crypto/hmac
+ * @private
+ */
+
+import { Hmac } from '@openpgp/asmcrypto.js/dist_es8/hmac/hmac';
+import { Sha1 } from '@openpgp/asmcrypto.js/dist_es8/hash/sha1/sha1';
+import { Sha256 } from '@openpgp/asmcrypto.js/dist_es8/hash/sha256/sha256';
+import { Sha512 } from '@openpgp/asmcrypto.js/dist_es8/hash/sha512/sha512';
```
Comment (twiss) [inline src/crypto/hmac.js]: asmcrypto.js's SHA512 implementation is very fast, but very big. It may not be worth it to import it just for hashing very small values very occasionally.
Tbh, perhaps you could reimplement HMAC, using `crypto.hash.digest`? If all the data is available at once, it should be quite simple, see https://en.wikipedia.org/wiki/HMAC#Implementation.

File: test/general/x25519.js [hunk 14]
```diff
@@ -237,7 +237,7 @@ module.exports = () => (openpgp.config.ci ? describe.skip : describe)('X25519 Cr
           expect(R).to.deep.eq(r);
           expect(S).to.deep.eq(s);
         }),
-        signature.verify(22, openpgp.enums.hash.sha256, { r: R, s: S }, publicParams, undefined, data).then(result => {
+        signature.verify(22, openpgp.enums.hash.sha256, { r: R, s: S }, publicParams, {}, undefined, data).then(result => {
```
Comment (twiss) [inline test/general/x25519.js]:
```suggestion
        signature.verify(22, openpgp.enums.hash.sha256, { r: R, s: S }, publicParams, null, undefined, data).then(result => {
```

**Review by twiss — COMMENTED** [review]
Some last nitpicks (hopefully :sweat_smile:)

File: src/openpgp.js [hunk 15]
```diff
@@ -69,9 +69,19 @@ export async function generateKey({ userIDs = [], passphrase = '', type = 'ecc',
...
+      if (type === 'symmetric') {
+        publicKey = null;
+      } else {
+        throw err;
+      }
+    }
     return {
       privateKey: formatObject(key, format, config),
-      publicKey: formatObject(key.toPublic(), format, config),
+      publicKey: publicKey,
```
Comment (twiss) [inline src/openpgp.js]: I know I kind of suggested the try/catch, but it shouldn't really be needed:
```suggestion
      publicKey: type !== 'symmetric' ? formatObject(key.toPublic(), format, config) : null,
```

**Review by twiss — COMMENTED** [review]
LGTM, one last nitpick:

**Review by twiss — APPROVED** [review]

**Comment by danrr** [comment]
🎊
